File: dotnet/CucumberExpressions/GroupBuilder.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Text.RegularExpressions;

namespace CucumberExpressions;

public class GroupBuilder {
    private readonly List<GroupBuilder> groupBuilders = new();
    private bool capturing = true;
    private String source;
    private int startIndex;
    private int endIndex;

    public GroupBuilder(int startIndex) {
        this.startIndex = startIndex;
    }

    public void add(GroupBuilder groupBuilder) {
        groupBuilders.Add(groupBuilder);
    }

    public Group build(Match matcher, IEnumerator<int> groupIndices)
    {
        groupIndices.MoveNext();
        int groupIndex = groupIndices.Current;
        var children = new List<Group>(groupBuilders.Count);
        foreach (GroupBuilder childGroupBuilder in groupBuilders) {
            children.Add(childGroupBuilder.build(matcher, groupIndices));
        }

        var matcherGroup = matcher.Groups[groupIndex];
        return new Group(matcherGroup.Success ? matcherGroup.Value : null, matcherGroup.Index, matcherGroup.Index + matcherGroup.Length, children);
    }

    public void setNonCapturing() {
        this.capturing = false;
    }

    public bool isCapturing() {
        return capturing;
    }

    public void moveChildrenTo(GroupBuilder groupBuilder) {
        foreach (GroupBuilder child in groupBuilders) {
            groupBuilder.add(child);
        }
    }

    public List<GroupBuilder> getChildren() {
        return groupBuilders;
    }

    public String getSource() {
        return source;
    }

    public void setSource(String source) {
        this.source = source;
    }

    public int getStartIndex() {
        return startIndex;
    }

    public int getEndIndex() {
        return endIndex;
    }

    public void setEndIndex(int endIndex) {
        this.endIndex = endIndex;
    }
}

File: python/cucumber_expressions/group_builder.py
from cucumber_expressions.group import Group


class GroupBuilder:
    def __init__(self):
        self.group_builders: list[GroupBuilder] = []
        self.capturing: bool = True
        self.source: str | None = None
        self.end_index: int | None = None

    def add(self, group_builder: "GroupBuilder"):
        self.group_builders.append(group_builder)

    def build(self, match, group_indices, group_name_map: dict) -> Group:
        group_index = next(group_indices)
        group_name = group_name_map.get(group_index, None)

        children = [
            gb.build(match, group_indices, group_name_map) for gb in self.group_builders
        ]
        return Group(
            name=group_name,
            value=match.group(group_index),
            start=match.regs[group_index][0],
            end=match.regs[group_index][1],
            children=children,
        )

    def move_children_to(self, group_builder: "GroupBuilder") -> None:
        for child in self.group_builders:
            group_builder.add(child)

    @property
    def children(self) -> list["GroupBuilder"]:
        return self.group_builders
